go/vt/vtgate/vindexes/region_json.go

```diff
@@ -0,0 +1,143 @@
+/*
+Copyright 2019 The Vitess Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package vindexes
+
+import (
+	"bytes"
+	"encoding/binary"
+	"encoding/json"
+	"io/ioutil"
+
+	"vitess.io/vitess/go/sqltypes"
+	"vitess.io/vitess/go/vt/key"
+	"vitess.io/vitess/go/vt/log"
+)
+
+var (
+	_ MultiColumn = (*RegionJson)(nil)
+)
+
+func init() {
+	Register("region_json", NewRegionJson)
+}
+
+// RegionMap is used to store mapping of country to region
+type RegionMap map[string]uint64
+
+// RegionJson defines a vindex that uses a lookup table.
+// The table is expected to define the id column as unique. It's
+// Unique and a Lookup.
+type RegionJson struct {
+	name        string
+	regionMap   RegionMap
+	regionBytes int
+}
+
+// NewRegionJson creates a RegionJson vindex.
+// The supplied map requires all the fields of "region_experimental".
+// Additionally, it requires a region_map argument representing the path to a json file
+// containing a map of country to region.
+func NewRegionJson(name string, m map[string]string) (Vindex, error) {
+	rmPath := m["region_map"]
+	rmap := make(map[string]uint64)
+	data, err := ioutil.ReadFile(rmPath)
+	if err != nil {
+		return nil, err
+	}
+	log.Infof("Loaded Region map from: %s", rmPath)
+	err = json.Unmarshal(data, &rmap)
+	if err != nil {
+		return nil, err
+	}
+
+	return &RegionJson{
+		name:      name,
+		regionMap: rmap,
+	}, nil
+}
+
+// String returns the name of the vindex.
+func (rv *RegionJson) String() string {
+	return rv.name
+}
+
+// Cost returns the cost of this index as 1.
+func (rv *RegionJson) Cost() int {
+	return 1
+}
+
+// IsUnique returns true since the Vindex is unique.
+func (rv *RegionJson) IsUnique() bool {
+	return true
+}
+
+// Map satisfies MultiColumn.
+func (rv *RegionJson) Map(vcursor VCursor, rowsColValues [][]sqltypes.Value) ([]key.Destination, error) {
+	destinations := make([]key.Destination, 0, len(rowsColValues))
+	for _, row := range rowsColValues {
+		if len(row) != 2 {
+			destinations = append(destinations, key.DestinationNone{})
+			continue
+		}
+		// Compute hash.
+		hn, err := sqltypes.ToUint64(row[0])
+		if err != nil {
+			destinations = append(destinations, key.DestinationNone{})
+			continue
+		}
+		h := vhash(hn)
+
+		log.Infof("Region map: %v", rv.regionMap)
```

Comment:
These log messages should be removed. They'd cause too much spam.

Reply:
@sougou Just those, or should the log message on line 61 also be removed?

Reply:
All log messages in `Map` should be removed, because it's a high QPS function,